image-builder/files_debootstrap/sbin/overlay-init

```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+# Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#       http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+set -x
+
+# Parameters:
+# 1. rw_root -- path where the read/write root is mounted
+# 2. work_dir -- path to the overlay workdir (must be on same filesystem as rw_root)
+# Overlay will be set up on /mnt, original root on /mnt/rom
+pivot() {
+    local rw_root work_dir
+    rw_root="$1"
+    work_dir="$2"
+    /bin/mount \
```

Comment:
Do the existing mounts under `/` propagate when we create the overlay at `/mnt`? In other words, when we make the overlay do we see e.g.`/mnt/overlay` and `/mnt/dev`? And if so, after we do the pivot_root do we see `/overlay` and `/dev`?

This is just for better understanding, no underlying concern

Reply:
When overlay-init runs, there are no other mounts. If needed, it could set some up, but for now it doesn't need any.

Any additional mounts that it creates will remain relative to the original root filesystem when we pivot. So, if we mount /proc, then pivot to /overlay with the original root moving to /rom, then the proc filesystem will move to /rom/proc. The real init process, executed at the end of overlay-init, would then be responsible for mounting the real /proc relative to the new root.

Reply:
Ah okay cool, so I'm guessing at this point of the boot process `/dev` is just filled with devices statically defined as part of the rootfs image and the typical devfs comes later from the real init.

overlay-init does create a mount at `/overlay` though earlier for the rw-dir and work-dir, so would that imply you can see `/rom/overlay` after overlay-init is done?

If so, I'm just imagining weirdness that could ensue, like after the pivot_root you could do `rm /foo/bar` and `bar` will be gone, but then if you do `rm /rom/overlay/root/foo/bar` you would delete the whiteout file and thus undelete the original `/foo/bar`.

If the above is accurate, maybe it's okay, but it might bring up some questions in more obscure use cases (like whether it's possible to securely delete a file from within the VM).

Reply:
Yeah, we should never be installing sensitive data in the image root. If we ever do, for some reason, the appropriate response would be to generate a replacement image root with the correct contents.